manim/mobject/text/code_mobject.py

```diff
@@ -89,549 +64,186 @@ def construct(self):
-        strati = self.html_string.find("background:")
-        self.background_color = self.html_string[strati + 12 : strati + 19]
-        self._gen_code_json()
+        if code_file is None and code_string is None:
+            raise ValueError("Either a code file or a code string must be specified.")
 
-        self.code = self._gen_colored_lines()
-        if self.insert_line_no:
-            self.line_numbers = self._gen_line_numbers()
-            self.line_numbers.next_to(self.code, direction=LEFT, buff=self.line_no_buff)
-        if self.background == "rectangle":
-            if self.insert_line_no:
-                foreground = VGroup(self.code, self.line_numbers)
-            else:
-                foreground = self.code
-            rect = SurroundingRectangle(
-                foreground,
-                buff=self.margin,
-                color=self.background_color,
-                fill_color=self.background_color,
-                stroke_width=self.background_stroke_width,
-                stroke_color=self.background_stroke_color,
-                fill_opacity=1,
-            )
-            rect.round_corners(self.corner_radius)
-            self.background_mobject = rect
+        if code_file is not None:
+            code_file = Path(code_file)
+            code_string = code_file.read_text(encoding="utf-8")
+            lexer = guess_lexer_for_filename(code_file.name, code_string)
         else:
-            if self.insert_line_no:
-                foreground = VGroup(self.code, self.line_numbers)
+            if language is not None:
+                lexer = get_lexer_by_name(language)
             else:
-                foreground = self.code
-            height = foreground.height + 0.1 * 3 + 2 * self.margin
-            width = foreground.width + 0.1 * 3 + 2 * self.margin
+                lexer = guess_lexer(code_string)
 
-            rect = RoundedRectangle(
-                corner_radius=self.corner_radius,
-                height=height,
-                width=width,
-                stroke_width=self.background_stroke_width,
-                stroke_color=self.background_stroke_color,
-                color=self.background_color,
-                fill_opacity=1,
+        formatter = HtmlFormatter(
+            style=formatter_style,
+            noclasses=True,
+            cssclasses="",
+        )
+        soup = BeautifulSoup(
+            highlight(code_string, lexer, formatter), features="html.parser"
+        )
+        self._code_html = soup.find("pre")
+
+        # as we are using Paragraph to render the text, we need to find the character indices
+        # of the segments of changed color in the HTML code
+        color_ranges = []
+        current_line_color_ranges = []
+        current_line_char_index = 0
+        for child in self._code_html.children:
+            if child.name == "span":
+                try:
+                    color = child["style"].removeprefix("color: ")
```

Comment:
`child` doesn't have a `name` attribute unless it's also a `Tag`, which is implicitly given. We should also assert that `child` is a `Tag`, either here or along with the previous assertion.

`child["style"]` can be either a string or a list of strings which does not have `.removeprefix()`. Maybe we should also assert that `child["style"]` is a `str`.

The code does get a bit clunky, so any suggestions are accepted.

```suggestion
        for child in self._code_html.children:
            assert isinstance(child, Tag)
            if child.name == "span":
                try:
                    style = child["style"]
                    assert isinstance(style, str)
                    color = style.removeprefix("color: ")
```

Reply:
`child` can also be `NavigableString`, the assertion will have to be rewritten slightly. Otherwise a good idea though, 👍

Reply:
I just checked, and indeed: it can be a `NavigableString`. So yeah, it would be `assert isinstance(child, (Tag, NavigableString))`.

Reply:
Actually, `self._code_html.children` is already typed as `Iterator[PageElement]`, that seems strong enough to me. I've strengthened the typing within the `try`-block in d110d5ec